File: 953/verifying-an-alien-dictionary.c
int comp(char c1, char c2, char* order){
    if(c1==c2) return 0; //0:相等
    for(int i=0; i<26; i++){
        if(order[i]==c1) return -1; //左邊比較前面
        if(order[i]==c2) return +1; //右邊比較前面
    }
    return 0;
}
int compStr(char* s1, char* s2, char* order){
    int i=0;
    for(   ; s1[i]!=0 && s2[i]!=0; i++){
        int diff = comp(s1[i],s2[i], order);
        if(diff!=0) return diff;
    }
    if(s1[i]==0 && s2[i]==0) return 0;
    if(s1[i]==0) return -1;
    else return +1;

}
bool isAlienSorted(char ** words, int wordsSize, char * order){
    //先把字母翻譯好,再看是否有排序好
    //不對不對, 應該說, 要有個 qsort()的comp()函式, 能幫忙比大小
    for(int i=0; i<wordsSize-1; i++){
        if(compStr(words[i], words[i+1], order)>0) return false;
    }
    return true;
}

File: 953/verifying-an-alien-dictionary.py
# 外星人用的奇怪字典，對應的a...z是不同的，要翻譯一下
# 如果 words 裡的word是照「外星人」的字母排序好，就True
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        alien = defaultdict(int)
        for i,c in enumerate(order): # 照著 order 建對照表
            alien[c] = i # 外星人的字母對照表，對應外星人的「字母序」
        for i in range(len(words)-1): # words[i] vs. words[i+1]
            for k,c in enumerate(words[i]):
                if k<len(words[i+1]): # 在長度範圍內
                    print(c, words[i+1][k])
                    if alien[c] > alien[words[i+1][k]]:
                        return False # 順序不合理，失敗
                    elif alien[c] < alien[words[i+1][k]]:
                        break # 合理，離開此迴圈，下一筆
                else: # 長度不合理，words[i]太長
                    return False # 前面都相同，但words[i]太長，也失敗
        return True # 檢測合格
